web/api/v1/api.go

```diff
@@ -469,13 +469,16 @@ func returnAPIError(err error) *apiError {
 		return nil
 	}
 
-	switch err.(type) {
-	case promql.ErrQueryCanceled:
-		return &apiError{errorCanceled, err}
-	case promql.ErrQueryTimeout:
-		return &apiError{errorTimeout, err}
-	case promql.ErrStorage:
-		return &apiError{errorInternal, err}
+	// The error may have been wrapped, so we should also check the parent ones.
+	for cause := err; cause != nil; cause = errors.Unwrap(cause) {
```

Comment:
Hm, do we need a loop?

Why not `errors.Cause`? This is the usual pattern and I don't we need unwrapped cause as second parameter :thinking: It's the code that matters.

The benefit of errors.Cause is to avoid this loop, which looks quite odd to me (first time I see something like this)

Reply:
Oh yes sure, done.